lib/Service/FederatedUserService.php

```diff
@@ -1160,4 +1184,47 @@ public function getGroupCircle(string $groupId): Circle {
 
 		return $circle;
 	}
+
+
+	/**
+	 * @param FederatedUser $federatedUser
+	 *
+	 * @return Circle
+	 * @throws SingleCircleNotFoundException
+	 */
+	private function getCachedSingleCircle(FederatedUser $federatedUser): Circle {
+		$key = $this->generateCacheKey($federatedUser);
+		if (!$this->cache->hasKey($key)) {
+			throw new SingleCircleNotFoundException();
+		}
+
+		try {
+			/** @var Circle $singleCircle */
+			$singleCircle = $this->deserializeJson($this->cache->get($key), Circle::class);
+		} catch (InvalidItemException $e) {
+			throw new SingleCircleNotFoundException();
+		}
+
+		return $singleCircle;
```

Comment:
hasKey is deprecated with the following note:
> deprecated 9.1.0 Directly read from GET to prevent race conditions

```suggestion
		$key = $this->generateCacheKey($federatedUser);
		$cachedData = $this->cache->get($key);
		if ($cachedData === null) {
			throw new SingleCircleNotFoundException();
		}

		try {
			/** @var Circle $singleCircle */
			$singleCircle = $this->deserializeJson($cachedData, Circle::class);
		} catch (InvalidItemException $e) {
			throw new SingleCircleNotFoundException();
		}

		return $singleCircle;
```

Also I don't see you purging the cache, so 15min sounds long. Or is it save to assume nothing ever fades from the table?

Reply:
Thank, I will use your code!

About the purge there is no reason for the data I am looking for to be updated from the table. Doing so will break thing.

However, just to avoid any issue in the future, I added a purge on userDeletedEvent.